pyriemann/utils/median.py

```diff
@@ -0,0 +1,158 @@
+"""Medians of SPD matrices."""
+
+import warnings
+import numpy as np
+
+from .base import sqrtm, invsqrtm, logm, expm
+from .distance import pairwise_distance
+from .mean import mean_euclid
+
+
+def median_euclid(X, tol=10e-6, maxiter=50, init=None):
+    r"""Euclidean geometric median of matrices.
+
+    The Euclidean geometric median minimizes the sum of Euclidean distances
+    :math:`d_E` to all matrices [1]_ [2]_:
+
+    .. math::
+        \arg \min_{\mathbf{M}} \sum_i d_E (\mathbf{M}, \mathbf{X}_i)
+
+    It is different from the marginal median provided by NumPy [3]_.
+
+    Parameters
+    ----------
+    X : ndarray, shape (n_matrices, n_channels, n_channels)
+        Set of matrices.
+    tol : float, default=10e-6
+        The tolerance to stop the gradient descent.
+    maxiter : int, default=50
+        The maximum number of iterations.
+    init : None | ndarray, shape (n_channels, n_channels), default=None
+        A matrix used to initialize the gradient descent.
+        If None, the Euclidean mean is used.
+
+    Returns
+    -------
+    M : ndarray, shape (n_channels, n_channels)
+        Euclidean geometric median.
+
+    Notes
+    -----
+    .. versionadded:: 0.3.1
+
+    References
+    ----------
+    .. [1] Weiszfeld E. "Sur le point pour lequel la somme des distances de n
+        points donnés est minimum", Tohoku Math J, 1937
+    .. [2] Vardi Y and Zhan C-H. "The multivariate L1-median and associated
+        data depth", PNAS, 2000
+    .. [3] https://numpy.org/doc/stable/reference/generated/numpy.median.html
+    """
+    if init is None:
+        M = mean_euclid(X)
+    else:
+        M = init
+
+    for _ in range(maxiter):
+        dists = pairwise_distance(
+            X,
+            M[np.newaxis, ...],
+            metric='euclid'
+        )[:, 0]
+        is_zero = (dists == 0)
+
+        inv_dists = 1 / dists[~is_zero]
+        weights = inv_dists / np.sum(inv_dists)
+        Mnew = np.einsum('a,abc->bc', weights, X[~is_zero])
```

Comment:
to me this is not a gradient descent. A gradient descent would look like:

Mnew = Mold - step_size * something

where is the step size in your code?

Reply:
You are right. Ostresh's implementation is a gradient descent, not Vardi's one.
I change the doc.